Shared read-only pages between TAs

jenswi-linaro opened the pull request:
This P-R introduces shared read-only pages between TAs. It actually also supports read/write pages as a side effect.

Note that  the patch "Replace ta_head.entry with elf entry" breaks backwards compatibility.

Some of the commits in this P-R can be a bit hard to review. I'm open to suggestions on how to proceed with the review.

jenswi-linaro commented:
Hopefully fixes the shippable error and the checkpatch errors/warnings that can be fixed.

jenswi-linaro commented:
Hopefully fixes the last shippable error and the checkpatch errors/warnings that can be fixed.

jenswi-linaro commented:
Now it's only the checkpatch warning about TAILQ_ENTRY()

jenswi-linaro commented:
Rebased on master

jenswi-linaro commented:
https://github.com/jenswi-linaro/optee_os/tree/map_temp_param
Here's a teaser with what's possible on top of this PR (allocating temporary secure memory directly in the TA instead of in the core). 
Allocating the heap directly in the TA instead is another possibility that I haven't explored yet.

jforissier commented:
> https://github.com/jenswi-linaro/optee_os/tree/map_temp_param
> Here's a teaser with what's possible on top of this PR

Looks pretty cool if you ask me :)

> Allocating the heap directly in the TA instead is another possibility that I haven't explored yet.

That would be awesome. Then we could get rid of a couple of ugly/hard-coded things such as https://github.com/OP-TEE/optee_os/blob/3.4.0/ta/arch/arm/user_ta_header.c#L64-L66 and https://github.com/OP-TEE/optee_os/blob/3.4.0/ta/arch/arm/ta.ld.S#L90-L103.

jenswi-linaro commented:
With "Replace ta_head.entry with elf entry" we're losing compatibility with TAs compiled with older OP-TEE releases. How big problem is this?
The problem is that we've forced an ELF section to be read-only and executable while we're still writing to it during relocation. We could add some copy-on-write feature to allow this behavior, but it would make things even more complicated and I'm not sure it will be needed for anything but slightly broken or sub-optimal ELFs.

jforissier commented:
> With "Replace ta_head.entry with elf entry" we're losing compatibility with TAs compiled with older OP-TEE releases.

Yeah, that is unfortunate.

> How big problem is this?

Frankly, I don't know. But as time passes it is becoming more likely that someone will want to run older TAs. Sometimes recompiling may be OK, but I can imagine that it may also be quite inconvenient (such as if third-parties are involved).

> The problem is that we've forced an ELF section to be read-only and executable while we're still writing to it during relocation.

That was a bad idea. Actually I wonder how I missed that, because there is a big fat comment screaming "BUG!" in `ta.ld.S` (here: https://github.com/OP-TEE/optee_os/blob/3.4.0/ta/arch/arm/ta.ld.S#L19-L24). We were simply not considering sharing pages at that time...

> We could add some copy-on-write feature to allow this behavior, but it would make things even more complicated and I'm not sure it will be needed for anything but slightly broken or sub-optimal ELFs.

How about we bypass the page sharing mechanism entirely when loading an "old" TA? It would be enough to ensure compatibility, and hopefully not too intrusive/complex?

jenswi-linaro commented:
> How about we bypass the page sharing mechanism entirely when loading an "old" TA? It would be enough to ensure compatibility, and hopefully not too intrusive/complex?

Good idea, I'll make an attempt at that.

jenswi-linaro commented:
Update with bypassing page sharing for old TAs.

jforissier commented:
I can reproduce a TA crash on HiKey960 when testing this PR with shared user-mode libraries, TA ASLR and pager (i.e., my branch: https://github.com/jforissier/optee_os/tree/shared-ro-pages--ta-aslr2 ; this is basically this PR merged with my TA ASLR PR, plus commit "Add TA entry point function: __ta_entry()").

```
$ export CFG_TEE_CORE_LOG_LEVEL=2
$ export CFG_ULIBS_SHARED=y
$ export CFG_WITH_PAGER=y
$ export CFG_TA_ASLR=y   # (note, enabled by default on HiKey960)
$ make 
[...]
root@HiKey960:/ xtest 4002& xtest 4002
Test ID: 4002
Run test suite with level=0

TEE test application started with device [(null)]
######################################################
#
# regression
#
######################################################

* regression_4002 Test TEE Internal API MAC operations
Test ID: 4002
Run test suite with level=0

TEE test application started with device [(null)]
######################################################
#
# regression
#
######################################################

* regression_4002 Test TEE Internal API MAC operations
o regression_4002.1 MAC case 0 algo 0x30000001
o regression_4002.1 MAC case 0 algo 0x30000001
  regression_4002.1 OK
  regression_4002.1 OK
o regression_4002.2 MAC case 1 algo 0x30000001
o regression_4002.2 MAC case 1 algo 0x30000001
  regression_4002.2 OK
  regression_4002.2 OK
o regression_4002.3 MAC case 2 algo 0x30000002
o regression_4002.3 MAC case 2 algo 0x30000002
  regression_4002.3 OK
  regression_4002.3 OK
o regression_4002.4 MAC case 3 algo 0x30000002
o regression_4002.4 MAC case 3 algo 0x30000002
  regression_4002.4 OK
  regression_4002.4 OK
o regression_4002.5 MAC case 4 algo 0x30000003
o regression_4002.5 MAC case 4 algo 0x30000003
  regression_4002.5 OK
o regression_4002.6 MAC case 5 algo 0x30000003
  regression_4002.6 OK
o regression_4002.7 MAC case 6 algo 0x30000004
  regression_4002.7 OK
o regression_4002.8 MAC case 7 algo 0x30000004
  regression_4002.8 OK
o regression_4002.9 MAC case 8 algo 0x30000004
  regression_4002.9 OK
o regression_4002.10 MAC case 9 algo 0x30000004
  regression_4002.10 OK
o regression_4002.11 MAC case 10 algo 0x30000005
  regression_4002.11 OK
o regression_4002.12 MAC case 11 algo 0x30000005
  regression_4002.12 OK
o regression_4002.13 MAC case 12 algo 0x30000006
  regression_4002.13 OK
o regression_4002.14 MAC case 13 algo 0x30000006
  regression_4002.14 OK
o regression_4002.15 MAC case 14 algo 0x30000110
  regression_4002.15 OK
o regression_4002.16 MAC case 15 algo 0x30000110
  regression_4002.16 OK
o regression_4002.17 MAC case 16 algo 0x30000111
  regression_4002.17 OK
o regression_4002.18 MAC case 17 algo 0x30000111
  regression_4002.18 OK
o regression_4002.19 MAC case 18 algo 0x30000113
  regression_4002.19 OK
o regression_4002.20 MAC case 19 algo 0x30000113
  regression_4002.20 OK
o regression_4002.21 MAC case 20 algo 0x30000510
  regression_4002.21 OK
o regression_4002.22 MAC case 21 algo 0x30000510
  regression_4002.22 OK
o regression_4002.23 MAC case 22 algo 0x30000511
  regression_4002.23 OK
o regression_4002.24 MAC case 23 algo 0x30000511
  regression_4002.24 OK
o regression_4002.25 MAC case 24 algo 0x30000513
  regression_4002.25 OK
o regression_4002.26 MAC case 25 algo 0x30000513
  regression_4002.26 OK
o regression_4002.27 MAC case 26 algo 0x30000510
  regression_4002.27 OK
o regression_4002.28 MAC case 27 algo 0x30000510
  regression_4002.28 OK
o regression_4002.29 MAC case 28 algo 0x30000511
  regression_4002.29 OK
o regression_4002.30 MAC case 29 algo 0x30000511
  regression_4002.30 OK
o regression_4002.31 MAC case 30 algo 0x30000513
  regression_4002.31 OK
o regression_4002.32 MAC case 31 algo 0x30000513
  regression_4002.32 OK
o regression_4002.33 MAC case 32 algo 0x30000513
  regression_4002.33 OK
o regression_4002.34 MAC case 33 algo 0x30000513
  regression_4002.34 OK
o regression_4002.35 MAC case 34 algo 0x30000610
  regression_4002.35 OK
o regression_4002.36 MAC case 35 algo 0x30000610
  regression_4002.36 OK
o regression_4002.37 MAC case 36 algo 0x30000610
  regression_4002.37 OK
o regression_4002.38 MAC case 37 algo 0x30000610
  regression_4002.38 OK
o regression_4002.39 MAC case 38 algo 0x30000610
  regression_4002.39 OK
o regression_4002.40 MAC case 39 algo 0x30000610
  regression_4002.40 OK
o regression_4002.41 MAC case 40 algo 0x30000610
  regression_4002.41 OK
o regression_4002.42 MAC case 41 algo 0x30000610
  regression_4002.42 OK
o regression_4002.43 MAC case 42 algo 0x30000610
  regression_4002.43 OK
o regression_4002.44 MAC case 43 algo 0x30000610
  regression_4002.44 OK
o regression_4002.45 MAC case 44 algo 0x30000610
  regression_4002.45 OK
o regression_4002.46 MAC case 45 algo 0x30000610
  regression_4002.46 OK
o regression_4002.47 MAC case 46 algo 0x30000610
  regression_4002.47 OK
o regression_4002.48 MAC case 47 algo 0x30000610
  regression_4002.48 OK
o regression_4002.49 MAC case 48 algo 0x30000610
  regression_4002.49 OK
o regression_4002.50 MAC case 49 algo 0x30000610
  regression_4002.50 OK
o regression_4002.51 MAC case 50 algo 0x30000610
  regression_4002.51 OK
o regression_4002.52 MAC case 51 algo 0x30000610
  regression_4002.52 OK
o regression_4002.53 MAC case 52 algo 0x30000610
  regression_4002.53 OK
o regression_4002.54 MAC case 53 algo 0x30000610
  regression_4002.54 OK
o regression_4002.55 MAC case 54 algo 0x30000610
  regression_4002.55 OK
  regression_4002 OK
+-----------------------------------------------------
Result of testsuite regression filtered by "4002":
regression_4002 OK
+-----------------------------------------------------
1290 subtests of which 0 failed
1 test case of which 0 failed
92 test cases were skipped
TEE test application done!
E/TC:5 1
E/TC:5 1 User TA data-abort at address 0x19 (translation fault)
E/TC:5 1  esr 0x92000045  ttbr0 0x200003f0260a0   ttbr1 0x00000000   cidr 0x0
E/TC:5 1  cpu #5          cpsr 0x20000130
E/TC:5 1  x0  00000000400a5c38 x1  0000000000000000
E/TC:5 1  x2  0000000000000020 x3  00000000400a5c38
E/TC:5 1  x4  0000000000000000 x5  000000004005ff48
E/TC:5 1  x6  00000000400a3988 x7  000000004005ff80
E/TC:5 1  x8  000000004005ff38 x9  00000000400728c8
E/TC:5 1  x10 0000000000000006 x11 0000000000000003
E/TC:5 1  x12 00000000400a3b08 x13 000000004005ff28
E/TC:5 1  x14 0000000040092a81 x15 0000000000000000
E/TC:5 1  x16 000000003f05f150 x17 0000000000000000
E/TC:5 1  x18 0000000000000000 x19 000000003f084d30
E/TC:5 1  x20 0000000000000001 x21 000000003f0128c0
E/TC:5 1  x22 00000000000005e0 x23 000000003f084f38
E/TC:5 1  x24 000000003f03b7b0 x25 000000003f03bcc0
E/TC:5 1  x26 000000003f084f00 x27 0000000000000006
E/TC:5 1  x28 000000004005ff88 x29 0000000000000000
E/TC:5 1  x30 000000003f04e33c elr 0000000040076278
E/TC:5 1  sp_el0 000000004005ff80
E/TC:5 1 Status of TA cb3e5ba0-adf1-11e0-998b-0002a5d5c51b (0x3f03bee0) (active)
E/TC:5 1  arch: arm  load address: 0x0000000040060000 ctx-idr: 2
E/TC:5 1  stack: 0x0000000040057000 36864
E/TC:5 1  region  0: va 0x0000000040000000 pa 0x000000003f002000 size 0x002000 flags ---R-X
E/TC:5 1  region  1: va 0x0000000040002000 pa 0x000000003f018000 size 0x001000 flags ---RW-
E/TC:5 1  region  2: va 0x0000000040057000 pa 0x0000000000000000 size 0x009000 flags rw-RW-
E/TC:5 1  region  3: va 0x0000000040060000 pa 0x0000000000000000 size 0x00a000 flags r-x--- [0]
E/TC:5 1  region  4: va 0x000000004006a000 pa 0x0000000000000000 size 0x009000 flags rw---- [0]
E/TC:5 1  region  5: va 0x0000000040073000 pa 0x0000000000000000 size 0x007000 flags r-x--- [1]
E/TC:5 1  region  6: va 0x000000004008a000 pa 0x0000000000000000 size 0x001000 flags rw---- [1]
E/TC:5 1  region  7: va 0x000000004008b000 pa 0x0000000000000000 size 0x009000 flags r-x--- [2]
E/TC:5 1  region  8: va 0x00000000400a3000 pa 0x0000000000000000 size 0x003000 flags rw---- [2]
E/TC:5 1  region  9: va 0x00000000400a6000 pa 0x0000000000000000 size 0x022000 flags r-x--- [3]
E/TC:5 1  region 10: va 0x00000000400d8000 pa 0x0000000000000000 size 0x004000 flags rw---- [3]
E/TC:5 1  region 11: va 0x00000000400dc000 pa 0x0000000000000000 size 0x002000 flags r-----
E/TC:5 1  [0] cb3e5ba0-adf1-11e0-998b-0002a5d5c51b @ 0x0000000040060000
E/TC:5 1  [1] 71855bba-6055-4293-a63f-b0963a737360 @ 0x0000000040073000
E/TC:5 1  [2] 527f1a47-b92c-4a74-95bd-72f19f4a6f74 @ 0x000000004008b000
E/TC:5 1  [3] 87bb6ae8-4b1d-49fe-9986-2b966132c309 @ 0x00000000400a6000
/home/jerome/work/optee_build_hikey960/optee_test/host/xtest/xtest_helpers.c:380: ret_orig has an unexpected value: 0x3 = TEEC_ORIGIN_TEE, expected 0x4 = TEEC_ORIGIN_TRUSTED_APP
/home/jerome/work/optee_build_hikey960/optee_test/host/xtest/regression_4000.c:1700: ta_crypt_cmd_free_operation(c, &session, op2) has an unexpected value: 0xffff3024 = TEE_ERROR_TARGET_DEAD, expected 0x0 = TEEC_SUCCESS
  regression_4002.5 FAILED
  regression_4002 FAILED
+-----------------------------------------------------
Result of testsuite regression filtered by "4002":
regression_4002.5 FAILED first error at /home/jerome/work/optee_build_hikey960/optee_test/host/xtest/xtest_helpers.c:380
regression_4002 FAILED
+-----------------------------------------------------
113 subtests of which 2 failed
1 test case of which 1 failed
92 test cases were skipped
TEE test application done!
[1]+  Done                       xtest 4002
root@HiKey960:/
```

The crash is not always the same (translation fault, alignment fault, write permission fault).

jenswi-linaro commented:
I'm not able to recreate the problem on my hikey 620. But I've found something suspicions. Does this patch help?
``` diff
--- a/core/arch/arm/mm/tee_pager.c
+++ b/core/arch/arm/mm/tee_pager.c
@@ -1502,6 +1502,7 @@ void tee_pager_pgt_save_and_release_entries(struct pgt *pgt)
 
                        area_set_entry(area, idx, 0, 0);
                        pgt_dec_used_entries(pgt);
+                       tlbi_mva_allasid(area_idx2va(area, idx));
                }
        }
        assert(!pgt->num_used_entries);

jforissier commented:
> Does this patch help?

Unfortunately, no :(

jforissier commented:
Apparently, when the first TA is unloaded it calls `tee_pager_rem_uta_areas()` and it is at this point that the second TA crashes.

etienne-lms commented on core/arch/arm/kernel/user_ta.c:
```diff
@@ -167,14 +173,12 @@ static TEE_Result get_elf_segments(struct user_ta_elf *elf,
-			segs[idx - 1].oend = MAX(segs[idx - 1].oend,
-						 segs[idx].oend);
+			segs[idx - 1].size = segs[idx].va + segs[idx].size -
+					     segs[idx - 1].va;
```
This assumes `segs[idx]` vmem range ends after `segs[idx -1]`.
Code below assumes `segs[idx]` vmem range **starts** after `segs[idx -1]` range.
Maybe few `assert()` would ensure those.

jenswi-linaro replied:
Agree, I'll fix.

etienne-lms commented on core/arch/arm/kernel/user_ta.c:
```diff
@@ -55,15 +61,21 @@ struct user_ta_elf {
 
+static void free_segs(struct load_seg *segs, size_t num_segs)
+{
+	for (size_t n = 0; n < num_segs; n++)
```
minor: local vars at function entry.

etienne-lms commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -484,7 +448,11 @@ KEEP_PAGER(area_insert_tail);
 {
+	size_t num_pages = size / SMALL_PAGE_SIZE;
 	struct tee_pager_area *area;
+	enum area_type at;
```
init value for these

etienne-lms commented on core/arch/arm/mm/mobj.c:
```diff
@@ -965,7 +967,9 @@ struct mobj *mobj_seccpy_shm_alloc(size_t size)
+				      SMALL_PAGE_SIZE);
+	if (tee_pager_add_uta_area(utc, va, m->fobj))
 		goto bad;
 
```
Add `fobj_put(m->fobj);` here.

jenswi-linaro replied:
You're right. Which commit is this?

etienne-lms replied:
In `core: change tee_pager_add_uta_area() arguments`, [here](https://github.com/OP-TEE/optee_os/pull/2801/commits/81c1ba1c9525b5d056f3adb1f7030e707c4409bb#diff-1b5dd72b044761b82ecce76d98e0bdf9R970).

jenswi-linaro replied:
No a `fobj_put()` isn't needed here, the pointer is saved at later put in `mobj_seccpy_shm_free()`

jforissier commented:
HiKey960 shared pages +pager+ASLR crash with some debug traces: https://pastebin.com/hCpRb7WE

It looks like the second instance of the TA crashes in the middle of the execution of both TAs, i.e., not when the first instance exits as I thought initially.

jforissier commented on ta/arch/arm/ta.ld.S:
```diff
@@ -53,7 +53,15 @@ SECTIONS {
         .ctors : { *(.ctors) }
         .dtors : { *(.dtors) }
-	.got : { *(.got.plt) *(.got) }
+	.dynamic : { *(.dynamic) } :dyn :rodata
```
If you move `.dynamic` to `rwdata` below (where it belongs), then all the custom flags in PHDR can be removed (and therefore the big ugly comment, too).

I suspect the linker script could be vastly simplified, but that's another topic.

etienne-lms commented on core/mm/file.c:
```diff
@@ -0,0 +1,173 @@
+	}
+	mutex_unlock(&file_mu);
+
+	if (did_remove)
```
maybe assert we found the reference. Caller should not provide a reference that do not match.

jenswi-linaro replied:
In the lines above we're assuming that it's a valid pointer if it's not `NULL` and we aren't searching for it either.

I see how it could be another safety measure to traverse the linked list of `struct file`, but then we wouldn't be talking about just a little assert.

jenswi-linaro commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -919,18 +930,18 @@ bool tee_pager_set_uta_area_attr(struct user_ta_ctx *utc, vaddr_t base,
-				assert(pa == get_pmem_pa(pmem));
-			else
-				pa = get_pmem_pa(pmem);
+			assert(pa == get_pmem_pa(pmem));
```
@etienne-lms it seems you're trying to comment on a specific commit here given that the diff is marked as "Outdated". Which commit is quite hard for me to tell even by searching through all diffs for something that matches.

jforissier replied:
For big PRs like this one, it is probably best to make one review per commit.

etienne-lms replied:
This was related to commit "core: pager: add pmem flags" where ([here](https://github.com/OP-TEE/optee_os/pull/2801/files#diff-5b29521d3e6f0086fdd5670fb29c63dcL1038)):
```patch
 			area_get_entry(pmem->area, pmem->pgidx, &pa, &a);
-			if (a & TEE_MATTR_VALID_BLOCK)
-				assert(pa == get_pmem_pa(pmem));
-			else
-				pa = get_pmem_pa(pmem);
+			assert(pa == get_pmem_pa(pmem));
```
Then commit "core: pager: don't store hidden page pa in table" removes the assertion:
```patch
 			area_get_entry(pmem->area, pmem->pgidx, &pa, &a);
-			assert(pa == get_pmem_pa(pmem));
```
In the end, this is fine since the phys address is no more read from `area_get_entry()` but from `get_pmem_pa(pmem)`. It is just the intermediate commits that maybe fail on `assert(pa == get_pmem_pa());` until it is removed by commit "core: pager: don't store hidden page pa in table".

jenswi-linaro replied:
This commit is mostly a transitional commit trying to take as small steps as possible in order to make bisect more useful.
I've run complete xtest with this commit in the past so I know it works. Why I thought it was good to leave the assert to finally remove it in "core: pager: don't store hidden page pa in table" I can't remember any longer.

etienne-lms commented on core/mm/file.c:
```diff
@@ -0,0 +1,173 @@
+	size_t n = 0;
+	int rc = memcmp(tag, f->tag, taglen);
+
+	for (n = taglen; !rc && n < sizeof(f->tag); n++)
```
This expects recorded tags do not end with zeros. I guess this layer does not want to put constraint on tags. Maybe store also to tag length.

jenswi-linaro replied:
Fair enough, the code will also be easier to read if I'm using an explicit taglen instead of just checking that the zero padding is where it's supposed to be. I'll fix.

etienne-lms commented on core/tee/tadb.c:
```diff
@@ -683,6 +683,20 @@ tee_tadb_ta_get_property(struct tee_tadb_ta_read *ta)
+		return TEE_ERROR_SHORT_BUFFER;
+	}
+
+	memcpy(tag, ta->entry.tag, sizeof(ta->entry.tag));
```
From comment of commit _"core: add tee_tadb_get_tag()"_:
> Adds the function tee_tadb_get_tag() which returns a tag that uniquely
identifies a TA.

Are we sure the tags generated in `tadb.c` will be unique per TA?

jenswi-linaro replied:
The tag from AES-GCM is as good as a hash.

etienne-lms commented on core/arch/arm/kernel/early_ta.c:
```diff
@@ -107,6 +109,32 @@ static TEE_Result early_ta_get_size(const struct user_ta_store_handle *h,
+
+	*tag_len = TEE_SHA256_HASH_SIZE;
+	if (!tag || *tag_len < TEE_SHA256_HASH_SIZE)
+		return TEE_ERROR_SHORT_BUFFER;
```
fix:
```
	if (!tag || *tag_len < TEE_SHA256_HASH_SIZE) {
		*tag_len = TEE_SHA256_HASH_SIZE;
		return TEE_ERROR_SHORT_BUFFER;
	}
	*tag_len = TEE_SHA256_HASH_SIZE;
```

jenswi-linaro replied:
Thanks, I managed to make that error a few times.

etienne-lms commented on core/arch/arm/kernel/elf_load.h:
```diff
@@ -30,6 +30,13 @@ struct user_ta_store_ops {
 			       size_t *size);
+
+	/*
+	 * Return the tag or hash of the TA binary. Used to uniqely
```
s/uniqely/uniquely/

etienne-lms commented on core/arch/arm/kernel/ree_fs_ta.c:
```diff
@@ -198,6 +198,20 @@ static TEE_Result ree_fs_ta_get_size(const struct user_ta_store_handle *h,
+
+	*tag_len = handle->shdr->hash_size;
+	if (!tag || *tag_len < handle->shdr->hash_size)
+		return TEE_ERROR_SHORT_BUFFER;
```
Fix:
```
	if (!tag || *tag_len < handle->shdr->hash_size) {
		*tag_len = handle->shdr->hash_size;
		return TEE_ERROR_SHORT_BUFFER;
	}
	*tag_len = handle->shdr->hash_size;
```

etienne-lms commented on core/arch/arm/kernel/ree_fs_ta.c:
```diff
@@ -350,20 +383,36 @@ static TEE_Result buf_ta_read(struct user_ta_store_handle *h, void *data,
+
+	*tag_len = handle->tag_len;
+	if (!tag || *tag_len < handle->tag_len)
+		return TEE_ERROR_SHORT_BUFFER;
```
fix
```
	if (!tag || *tag_len < handle->tag_len) {
		*tag_len = handle->tag_len;
		return TEE_ERROR_SHORT_BUFFER;
	}
	*tag_len = handle->tag_len;
```

etienne-lms commented on core/tee/tadb.c:
```diff
@@ -686,13 +686,11 @@ tee_tadb_ta_get_property(struct tee_tadb_ta_read *ta)
-	if (*tag_len < sizeof(ta->entry.tag)) {
-		*tag_len = sizeof(ta->entry.tag);
+	*tag_len = sizeof(ta->entry.tag);
+	if (!tag || *tag_len < sizeof(ta->entry.tag))
```
fix
```
	if (!tag || *tag_len < sizeof(ta->entry.tag)) {
		*tag_len = sizeof(ta->entry.tag);
		return TEE_ERROR_SHORT_BUFFER;
	}
	*tag_len = sizeof(ta->entry.tag);
```

jenswi-linaro commented on core/arch/arm/mm/mobj.c:
```diff
@@ -1002,17 +1002,61 @@ static struct fobj *mobj_with_fobj_get_fobj(struct mobj *mobj)
+	if (!cattr)
+		return TEE_ERROR_GENERIC;
+
+	*cattr = TEE_MATTR_CACHE_CACHED;
```
I'll add a comment.

jenswi-linaro commented on core/arch/arm/kernel/user_ta.c:
```diff
@@ -455,7 +459,8 @@ static void free_utc(struct user_ta_ctx *utc)
 	tee_pager_rem_uta_areas(utc);
 	TAILQ_FOREACH(elf, &utc->elfs, link)
-		release_ta_memory_by_mobj(elf->mobj_code);
+		for (size_t n = 0; n < elf->num_segs; n++)
```
I'll fix.

jenswi-linaro commented on core/arch/arm/mm/mobj.c:
```diff
@@ -978,10 +978,74 @@ struct mobj *mobj_seccpy_shm_alloc(size_t size)
+{
+	assert(to_mobj_with_fobj(mobj));
+
+	return attr == CORE_MEM_SEC || attr == CORE_MEM_TEE_RAM;
```
I'm dropping `CORE_MEM_TEE_RAM` and only classify it as `CORE_MEM_SEC` which is needed to map it properly.

jenswi-linaro commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -919,18 +930,18 @@ bool tee_pager_set_uta_area_attr(struct user_ta_ctx *utc, vaddr_t base,
 			if (a == f)
 				continue;
 			area_set_entry(pmem->area, pmem->pgidx, 0, 0);
 			tlbi_mva_allasid(area_idx2va(pmem->area, pmem->pgidx));
```
There's no user read-only pages, all are starting out as read-write pages when loading the ELF. At the end they are remapped as read-only. Some pages may still be dirty but that's OK, the state is recorded.

jenswi-linaro commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -1078,19 +1164,36 @@ static struct tee_pager_pmem *tee_pager_get_page(struct tee_pager_area *area)
+
+		TAILQ_FOREACH(area, &pmem->fobj->areas, fobj_link) {
+			if (!area->pgt ||
+			    pmem->fobj_pgidx < area->fobj_pgoffs ||
```
OK, I'll refactor this. By the way, I think the omitted TLB invalidation is a bug.

jenswi-linaro commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -1397,21 +1506,29 @@ void tee_pager_add_pages(vaddr_t vaddr, size_t npages, bool unmap)
+			area = find_area(&tee_pager_area_head, va);
+			assert(area);
+			pmem->fobj = area->fobj;
+			assert(area->pgt == find_core_pgt(va));
```
OK, I'll combine it with the `assert(area)` above.

jenswi-linaro commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -968,25 +1012,49 @@ bool tee_pager_set_uta_area_attr(struct user_ta_ctx *utc, vaddr_t base,
 KEEP_PAGER(tee_pager_set_uta_area_attr);
 #endif /*CFG_PAGED_USER_TA*/
 
+void tee_pager_invalidate_fobj(struct fobj *fobj)
```
I prefer `tee_pager_invalidate_fobj()` especially if you look where it's called.

jenswi-linaro commented on core/mm/file.c:
```diff
@@ -0,0 +1,173 @@
+			unsigned int taglen)
+{
+	size_t n = 0;
+	int rc = memcmp(tag, f->tag, taglen);
```
I'll add

jenswi-linaro commented on core/mm/file.c:
```diff
@@ -0,0 +1,173 @@
+	if (!f)
+		return NULL;
+
+	memcpy(f->tag, tag, taglen);
```
I'll add a test

jenswi-linaro commented:
I've addressed the comments. Please check that you're OK with the changes. After that I'll squash them into their respective base commit and apply eventual tags. I'll also remove the commit "core: remove flags argument from tee_pager_alloc()" since it's quite bad to have the bignum memory pool locked into memory.

jenswi-linaro commented:
Squashed in review fixups and applied one tag.

jenswi-linaro commented:
Rebased on master.
@etienne-lms , what do you want me to do with the LGTMs? Is it R-B tags?

etienne-lms commented on core/mm/fobj.c:
```diff
@@ -0,0 +1,305 @@
+	assert(refcount_val(&fobj->refc));
+	assert(page_idx < fobj->num_pages);
+
+	if (!state->iv) {
```
maybe add a comment here:
`/* No IV yet, init brand new rw page */`

etienne-lms commented on core/mm/fobj.c:
```diff
@@ -0,0 +1,305 @@
+	uint8_t *dst = rwp->store + page_idx * SMALL_PAGE_SIZE;
+	struct rwp_aes_gcm_iv iv = {
+		{ (vaddr_t)state, state->iv >> 32, state->iv }
+	};
```
move the definitions to function entry?

jenswi-linaro replied:
I'll fix

etienne-lms commented on core/arch/arm/kernel/user_ta.c:
```diff
@@ -168,14 +176,22 @@ static TEE_Result get_elf_segments(struct user_ta_elf *elf,
+			size_t size = 0;
+
+			/*
+			 * All segments are supposted to be in order and if
```
s/supposted/supposed/

etienne-lms commented on core/arch/arm/mm/mobj.c:
```diff
@@ -978,10 +978,74 @@ struct mobj *mobj_seccpy_shm_alloc(size_t size)
+	if (!fobj)
+		return NULL;
+
+	struct mobj_with_fobj *m = calloc(1, sizeof(*m));
```
@jforissier, are you ok with such local var definition? (not at function entry).

jenswi-linaro replied:
I'm sorry, this is leftovers from before when it was OK. I'll fix.

etienne-lms commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -622,12 +616,14 @@ static void tee_pager_save_page(struct tee_pager_pmem *pmem, uint32_t attr)
 		size_t offs = pmem->area->base & CORE_MMU_PGDIR_MASK;
-		size_t idx = pmem->pgidx - (offs >> SMALL_PAGE_SHIFT);
+		size_t fobj_pgidx = (pmem->pgidx - (offs >> SMALL_PAGE_SHIFT)) +
+				     pmem->area->fobj_pgidx;
```
minor: fix indentation: 1 extra space here.

etienne-lms replied:
Outdated change: feel free to ignore my comment.

jenswi-linaro replied:
I'll fix

etienne-lms commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -645,8 +641,9 @@ static bool pager_add_uta_area(struct user_ta_ctx *utc, vaddr_t base,
 	struct tee_pager_area *area;
-	uint32_t flags;
 	vaddr_t b = base;
+	struct fobj *fobj;
```
`= NULL`

etienne-lms commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -445,44 +439,36 @@ static void area_insert_tail(struct tee_pager_area *area)
 	struct tee_pager_area *area;
-	enum area_type at;
-	struct fobj *fobj;
+	uint32_t flags;
```
`= 0`

etienne-lms commented on core/arch/arm/include/mm/tee_pager.h:
```diff
@@ -87,22 +87,23 @@ void tee_pager_add_core_area(vaddr_t base, enum tee_pager_area_type type,
  * finalized with tee_pager_set_uta_area_attr() to get more strict settings.
  *
- * Return true on success of false if the area can't be added
+ * Return TEE_SUCCESS on success,  anything else if the area can't be added
```
1 extra scape before `anything`

etienne-lms commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -673,11 +660,13 @@ static bool pager_add_uta_area(struct user_ta_ctx *utc, vaddr_t base,
+TEE_Result tee_pager_add_uta_area(struct user_ta_ctx *utc, vaddr_t base,
+			    struct fobj *fobj)
 {
+	TEE_Result res;
```
`= TEE_SUCCESS`

etienne-lms commented on core/arch/arm/mm/mobj.c:
```diff
@@ -1002,17 +1008,61 @@ static struct fobj *mobj_with_fobj_get_fobj(struct mobj *mobj)
+					size_t granule, paddr_t *pa)
+{
+	struct mobj_with_fobj *f = to_mobj_with_fobj(mobj);
+	paddr_t p;
```
`= 0`

etienne-lms commented on core/include/mm/fobj.h:
```diff
@@ -112,6 +114,29 @@ static inline TEE_Result fobj_save_page(struct fobj *fobj,
+ * Returns a valid pointer on success or NULL on failure.
+ */
+#ifdef CFG_PAGED_USER_TA
+#define fobj_ta_mem_alloc(num_pages) fobj_rw_paged_alloc((num_pages))
```
Tab between macro and its value. Maybe s/`((num_pages))`/`(num_pages)`/.

Comment applies also to line 137.

etienne-lms commented on core/mm/fobj.c:
```diff
@@ -303,3 +303,78 @@ static struct fobj_ops ops_locked_paged __rodata_unpaged = {
+struct fobj *fobj_sec_mem_alloc(unsigned int num_pages)
+{
+	struct fobj_sec_mem *f = calloc(1, sizeof(*f));
+	size_t size;
```
`= 0`

etienne-lms commented on core/mm/fobj.c:
```diff
@@ -303,3 +303,78 @@ static struct fobj_ops ops_locked_paged __rodata_unpaged = {
+	if (!f->mm)
+		goto err;
+
+	void *va = phys_to_virt(tee_mm_get_smem(f->mm), MEM_AREA_TA_RAM);
```
define local variable at function entry

etienne-lms commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -983,51 +983,41 @@ static struct tee_pager_pmem *pmem_find(struct tee_pager_area *area,
+	/*
+	 * If it's not a dirty block, then it should be
+	 * read only.
+	 */
```
minor: single line comment here

etienne-lms commented on core/arch/arm/include/mm/pgt_cache.h:
```diff
@@ -17,6 +17,7 @@
 #include <sys/queue.h>
 #include <types_ext.h>
 #include <util.h>
+#include <assert.h>
```
move to top

etienne-lms commented:
commit "core: pager: add missing tzsram physical pages": 
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

commit "core: pager: fix tbl_usage_count()": 
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

commit "core: pager: record fobj in pmem instead of area": 
Some comments, all minor...

etienne-lms commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -924,30 +989,35 @@ bool tee_pager_set_uta_area_attr(struct user_ta_ctx *utc, vaddr_t base,
 			dsb_ishst();
 
+			/*
+			 * Here's a problem if this page already is shared.
```
Maybe rephrase or elaborate.

etienne-lms commented:
commit "core: pager: record fobj in pmem instead of area":
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` with minor comments.

commit "core: add get_fobj() to mobj_seccpy_shm":
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

commit "core: user_ta: support mapping paged parameters":
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` with minor comments.

commit "core: remove unused pgt_transfer()":
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`

commit "core: remove unused tee_pager_transfer_uta_region()":
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`

etienne-lms commented:
commit "core: add prot arg to tee_pager_add_uta_area()":
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` (minor comments)

etienne-lms commented on core/include/mm/file.h:
```diff
@@ -0,0 +1,79 @@
+#define FILE_TAG_SIZE	TEE_SHA256_HASH_SIZE
+
+/*
+ * struct file_slice - one slices of a file
```
s/slices/slice/

etienne-lms commented on core/mm/file.c:
```diff
@@ -0,0 +1,168 @@
+		f->slices[n].fobj = fobj_get(slices[n].fobj);
+		if (!f->slices[n].fobj ||
+		    ADD_OVERFLOW(slices[n].page_offset,
+				 slices[n].fobj->num_pages, &s))
```
For sanity, maybe also check there is not already a matching slice, something like `assert(!file_find_slice(f, slices[n].page_offset))` before the slice is stored.

jenswi-linaro replied:
This became a bit complicated and also highlight an error, I'll take this as a `[review]` patch on top.

etienne-lms commented on core/include/mm/file.h:
```diff
@@ -0,0 +1,79 @@
+ * @f:		 File pointer
+ * @page_offset: Offset that must be covered
+ *
+ * If a matching file slices is found it's returned, else NULL is returned.
```
s/slices/slice/

etienne-lms commented on core/arch/arm/kernel/user_ta.c:
```diff
@@ -219,6 +222,38 @@ static struct mobj *alloc_ta_mem(size_t size)
+	 * this function. Since file currently exists all its fobjs are
+	 * also guaranteed to exist here.
+	 *
+	 * If mobj_with_fobj_alloc()
```
cosmetic: unwrap line here.

etienne-lms commented on core/arch/arm/kernel/user_ta.c:
```diff
@@ -219,6 +222,38 @@ static struct mobj *alloc_ta_mem(size_t size)
+	 * If a fobj is found it has to match with the start or something is
+	 * wrong, besides we wouldn't be able to map it properly either.
+	 */
+	if (fs->page_offset != page_offset)
```
maybe add `assert(fs->page_offset == page_offset);` for debug...

etienne-lms commented on core/arch/arm/kernel/user_ta.c:
```diff
@@ -712,6 +747,58 @@ static TEE_Result add_deps(struct user_ta_ctx *utc __unused,
+	struct file_slice *fs = NULL;
+	size_t num_slices = 0;
+	size_t n = 0;
+	uint8_t tag[FILE_TAG_SIZE];
```
I guess here it is fine to not init `tag`. But that's against the rule.

jenswi-linaro replied:
I'll fix

etienne-lms commented on core/arch/arm/kernel/user_ta.c:
```diff
@@ -879,12 +988,21 @@ static TEE_Result load_elf_from_store(const TEE_UUID *uuid,
 	if (res)
 		goto out;
 
+	if (!file && ta_head->depr_entry == UINT64_MAX) {
```
Maybe a word here about legacy TA support that mandates no page sharing.

etienne-lms commented:
commit "core: add tee_tadb_get_tag()":
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

commit "core: add get_tag() to struct user_ta_store_ops":
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

commit "core: share sections of loaded elf":
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` with minor comments.

jenswi-linaro commented on core/mm/fobj.c:
```diff
@@ -0,0 +1,305 @@
+	struct rwp_state *state = rwp->state + page_idx;
+	uint8_t *src = rwp->store + page_idx * SMALL_PAGE_SIZE;
+	struct rwp_aes_gcm_iv iv = {
+		{ (vaddr_t)state, state->iv >> 32, state->iv }
```
I'll fix

jenswi-linaro commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -1045,19 +1139,24 @@ static bool tee_pager_release_one_phys(struct tee_pager_area *area,
-		area_set_entry(area, pgidx, 0, 0);
+		assert(TAILQ_FIRST(&pmem->fobj->areas) == area);
+		assert(TAILQ_LAST(&pmem->fobj->areas,
+				  tee_pager_area_head) == area);
```
It's nice to have, I'll add a comment explaining that we're asserting that the pmem isn't shared.

jenswi-linaro commented:
Thanks to some of the comments by @etienne-lms I've discovered some problems while testing. Once understood and fixed I'll push an update.

jenswi-linaro commented:
I've addressed all comments. There's now a few `[review]` commits that I'll use to fix up the original commit with in a day or two.

The original branch needed rebasing and I think I've managed to stabilize things now with a few extra commits. Especially "core: tee_ta_init_user_ta_session(): flush pgt on error" and "core: invalidate entire icache" fixes tricky problems.

This branch obviously needs to be rebased again, I'll do that when fixing up all the commit with the `[reviewed]` commits.

jenswi-linaro commented:
Replaced by https://github.com/OP-TEE/optee_os/pull/2971
